patches/api/0420-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,746 @@
+new file mode 100644
+index 0000000000000000000000000000000000000000..196dde20685b07ee699c5b50330b1f9596082056
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
+@@ -0,0 +1,25 @@
++package io.papermc.paper.command.brigadier;
++
++import com.mojang.brigadier.Message;
++import net.kyori.adventure.text.Component;
++import net.kyori.adventure.text.serializer.ComponentSerializer;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.ServiceLoader;
++
++/**
++ * A component serializer that converts it to a brigadier compatible type.
++ */
++public interface MessageComponentSerializer extends ComponentSerializer<Component, Component, Message> {
++
++    /**
++     * A component serializer that converts it to a brigadier compatible type.
++     *
++     * @return serializer instance
++     */
++    @NotNull
++    static MessageComponentSerializer message() {
++        return MessageComponentSerializerHolder.PROVIDER;
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..f3b8cc535cfa74a16cd35ed5677c6958b894046d
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java
+@@ -0,0 +1,10 @@
++package io.papermc.paper.command.brigadier;
++
++import java.util.ServiceLoader;
++
++final class MessageComponentSerializerHolder {
++
++    public static final MessageComponentSerializer PROVIDER = ServiceLoader.load(MessageComponentSerializer.class)
++        .findFirst()
++        .orElseThrow();
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java b/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..14ff48fb772a9f0ca9a3d37f8a600a39d1d14a8c
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java
+@@ -0,0 +1,24 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import io.papermc.paper.command.brigadier.CommandSourceStack;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Represents an argument which requires
++ * the context of a command source stack to be fully resolved.
++ * @param <T> type
++ */
++@ApiStatus.Experimental
++public interface ArgumentResolver<T> {
++
++    /**
++     * Resolves the argument with the given
++     * command source stack.
++     * @param sourceStack source stack
++     * @return resolved
++     */
++    @NotNull
++    T resolve(@NotNull CommandSourceStack sourceStack) throws CommandSyntaxException;
```

Comment:
Is it smart to have this method marked as `@NotNull` when in the future, I feel like there could be a valid null to return here?

Reply:
Changed, do agree here.